SMLHelper/Patchers/EnumPatcher.cs

```diff
@@ -0,0 +1,113 @@
+﻿namespace SMLHelper.V2.Patchers
+{
+    using System;
+    using System.Collections.Generic;
+    using Harmony;
+    using SMLHelper.V2.Utility;
+
+    internal class EnumPatcher
+    {
+        internal static void Patch(HarmonyInstance harmony)
+        {
+            PatchUtils.PatchClass(harmony);
+
+            Logger.Log("EnumPatcher is done.", LogLevel.Debug);
+        }
+
+        [PatchUtils.Postfix]
+        [HarmonyPatch(typeof(Enum), nameof(Enum.GetValues))]
+        private static void Postfix_GetValues(Type enumType, ref Array __result)
+        {
+            if (enumType == typeof(TechType))
+            {
+                __result = GetValues(TechTypePatcher.cacheManager, __result);
+            }
+            else if (enumType == typeof(CraftTree.Type))
+            {
+                __result = GetValues(CraftTreeTypePatcher.cacheManager, __result);
+            }
+            else if (enumType == typeof(PingType))
+            {
+                __result = GetValues(PingTypePatcher.cacheManager, __result);
+            }
+        }
+
+        private static T[] GetValues<T>(EnumCacheManager<T> cacheManager, Array __result) where T : Enum
+        {
+            var list = new List<T>();
+            foreach (T type in __result)
+            {
+                list.Add(type);
+            }
+
+            list.AddRange(cacheManager.ModdedKeys);
+            return list.ToArray();
+        }
+
+        [PatchUtils.Prefix]
+        [HarmonyPatch(typeof(Enum), nameof(Enum.IsDefined))]
+        private static bool Prefix_IsDefined(Type enumType, object value, ref bool __result)
+        {
+            if (IsDefined(TechTypePatcher.cacheManager, enumType, value) ||
+                IsDefined(CraftTreeTypePatcher.cacheManager, enumType, value) ||
+                IsDefined(PingTypePatcher.cacheManager, enumType, value))
```

Comment:
It's not about `PatchUtils`, but maybe put these three managers into some dictionary by enum type? They can even add themselves to `EnumPatcher` in `EnumCacheManager` constructor, for example, so there will be no concrete enum types in `EnumPatcher`.

Reply:
I thought about that. I decided to keep it simple for this first iteration and return to it in the event that we ever add a 4th enum to be patched.

Reply:
Well, I think even with two enum types, more general approach would be useful in this case. But anyway, one `EnumPatcher` is better than having three similar patchers :)

Reply:
While I do agree that a general approach would be the most helpful, it would require abstracting the 3 original enum patching classes in a way they could be used without an explicit reference to the specific enum type they handle.  
The generics involved make this harder than it looks since each of 3 `EnumCacheManager<T>` references are actually completely different types.

In short, it would be a lot of over engineering for not only little gain in terms of code readability but also would add additional overhead to these enum patch methods.

Reply:
Ah, for some reason I thought that something like `List<EnumCacheManager<Enum>>` is possible. Yeah, you right, it's not worth it then.